### PR TITLE
[DEV APPROVED] TP#8070: Adds ability to download registered advisers as a CSV file

### PR DESCRIPTION
**Pull request by Tr4pSt3R**
[Card TP#8070: Registered Advisers Export List](https://moneyadviceservice.tpondemand.com/entity/8070)

yj6onbvdbp

### CODE REVIEW

File: app/controllers/admin/reports/registered_adviser_controller.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#  def self.to_csv(_ = {})
+#    CSV.generate do |csv|
+#      csv << ['Ref. Number', 'Name', 'Firm']
+
+#      @advisers.each do |adviser|
+#        csv << [adviser.reference_number, adviser.name, adviser.firm.registered_name]
+#      end
+#    end
+#  end
+# end
```
Comment (margOnline) [inline app/controllers/admin/reports/registered_adviser_controller.rb]: Can all this commented out code be removed.

File: app/views/admin/reports/registered_adviser/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        <td><%= link_to adviser.name, admin_adviser_path(adviser) %></td>
+        <td><%= link_to adviser.firm.registered_name, admin_firm_path(adviser.firm) %></td>
+      </tr>
+    <% end %>
+  <% else %>
+    <tr>
+      <td colspan="3">No advisers to display</td>
+    </tr>
+  <% end %>
+	<%= paginate @registered_advisers %>
```
Comment (margOnline) [inline app/views/admin/reports/registered_adviser/index.html.erb]: fix spacing

File: lib/ext/adviser.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+module AdviserExtension
+  extend ActiveSupport::Concern
+
+  class_methods do
+    def to_csv(_ = {})
+      CSV.generate do |csv|
+        csv << ['Ref. Number', 'Name', 'Firm']
+
+        find_each do |adviser|
```
Comment (margOnline) [inline lib/ext/adviser.rb]: 👍  on find_each

File: spec/features/admin/export_advisers_list_spec.rb [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+# The criteria for entry to the Money Advice Service directory was decided by
+# an independent panel consisting of both consumer and advisers representative
+# bodies.
+# We recently found that some of this advisers have been entering our
+# directory through other company that they are creating. These one are
+# potentialy scams.
+#
+# In order to make sure we don't have any of these advisers in our directory
+# (RAD). Therefore, we need to compare the list of advisers registered under
+# St James place on the FCA register with our database.
```
Comment (margOnline) [inline spec/features/admin/export_advisers_list_spec.rb]: I dont think this is the appropriate place for this information. Your tests are nicely worded and explain what's going on well enough.

File: app/controllers/admin/reports/registered_adviser_controller.rb (hunk 1 above)
Comment (tomas-stefano) [inline app/controllers/admin/reports/registered_adviser_controller.rb]: You can remove this comment lines here.

File: db/schema.rb [hunk 5]
```diff
@@ -15,7 +15,6 @@
 
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
-  enable_extension "pg_stat_statements"
```
Comment (tomas-stefano) [inline db/schema.rb]: Is this was intended for the csv to work?

File: lib/ext/adviser.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+          csv << [adviser.reference_number, adviser.name, adviser.firm.registered_name]
+        end
+      end
+    end
+  end
+end
+
+ActiveRecord::Base.send(:include, AdviserExtension)
```
Comment (tomas-stefano) [inline lib/ext/adviser.rb]: Is there any more benefits to be a module rather than a class that wraps the advisers list?

File: config/initializers/csv_renderer.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+require 'ext/adviser'
+
+ActionController.add_renderer :csv do |csv, options|
+  filename = options[:filename] || options[:template]
```
Comment (margOnline) [inline config/initializers/csv_renderer.rb]: Would be nice for the filename to include a timestamp.

File: lib/ext/adviser.rb (hunk 6 above)
Comment (tomas-stefano) [inline lib/ext/adviser.rb]: If you're using this module only in one class, why include in all active record base? This means that if I call Office.to_csv they assume that the object quacks like an Adviser. I think that's was not the intended.

Comment (tomas-stefano) [inline lib/ext/adviser.rb]: I would add a unit tests too around the to_csv.

File: app/views/admin/reports/registered_adviser/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+</p>
+
+<p>
+  The <a href="https://register.fca.org.uk/">Financial Services Register</a>
+  can be searched to determine exactly what status change has occurred.
+</p>
+
+<% if @registered_advisers.present? %>
+  <p>
+    <%= link_to 'Download as CSV', admin_reports_registered_adviser_index_path(format: :csv), download: 'registered_advisers.csv', class: 'btn btn-primary' %>
```
Comment (margOnline) [inline app/views/admin/reports/registered_adviser/index.html.erb]: I'd like to see us move away from long lines, ideally to 80 chars, but this is open to agreement by the team.

**Comment by tomas-stefano** [comment]
I would reword or squash the commits too.

File: app/controllers/admin/reports/registered_adviser_controller.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+module Admin
+  module Reports
+    class RegisteredAdviserController < Admin::ApplicationController
+      respond_to :html, :csv
+
+      def index
+        @registered_advisers = Adviser.all.page(params[:page])
+        respond_with(@registered_advisers)
```
Comment (willhall88) [inline app/controllers/admin/reports/registered_adviser_controller.rb]: It looks to me that `@registered_advisers` is a paginated number of the total. This is fine for the html page, *but* does the csv then only return the same results rather than ALL of the registered advisors? 
Is this the requirements we need and are there any tests to cover this count?

File: spec/models/adviser_spec.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+RSpec.describe Adviser, type: :model do
```
Comment (willhall88) [inline spec/models/adviser_spec.rb]: Perhaps this test should be updated to test `lib/ext/adviser.rb` and more explicitly test the outcome of that method.

File: lib/ext/adviser.rb (hunk 6 above)
Comment (willhall88) [inline lib/ext/adviser.rb]: Agree with @tomas-stefano, the test - [here](https://github.com/moneyadviceservice/rad/pull/406/files#diff-fcb272b2a5bcd68e29fa8e3e4051789dR1) should probably be updated to more explicitly test this code

File: spec/models/adviser_spec.rb (hunk 10 above)
Comment (willhall88) [inline spec/models/adviser_spec.rb]: initial thoughts ...

move `spec/models/adviser_spec.rb` -> `spec/lib/adviser_spec.rb`

- `Advisor` class should respond to csv
- `Office` class should not
- Response of `.to_csv` should return a csv with the headers `'Ref. Number', 'Name', 'Firm'`

Comment (tomas-stefano) [inline spec/models/adviser_spec.rb]: What's the difference between the Adviser model and the Advisor in the lib?

Comment (willhall88) [inline spec/models/adviser_spec.rb]: Adviser was originally put in `models/` but then made better sense to have it in `lib/`, just looks like the spec file hasn't been updated accordingly.

File: config/initializers/csv_renderer.rb (hunk 7 above)
Comment (willhall88) [inline config/initializers/csv_renderer.rb]: ♻️  @margOnline : https://github.com/moneyadviceservice/rad/pull/406/commits/65ba592e05deb5fac5a37900d5cff6c53428d92d

File: spec/controllers/admin/reports/registered_adviser_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
+RSpec.describe Admin::Reports::RegisteredAdviserController, type: :controller do
+  describe '#filename' do
```
Comment (tomas-stefano) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: IMHO, the only tests that you should worry about are the ones testing the public interface. Otherwise you may be coupling too tightly to implementation. Or make it the filename method to be public. Or even better, you can move the timestamp call to the renderer object and then your acceptance tests cover it.

File: lib/ext/adviser.rb [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+        find_each do |adviser|
+          csv << [adviser.reference_number, adviser.name, adviser.firm.registered_name]
+        end
+      end
+    end
+  end
+end
+
+Adviser.send(:include, AdviserExtension)
```
Comment (tomas-stefano) [inline lib/ext/adviser.rb]: Is there a reason to add the include outside of the Adviser class definitions?
Other thing, I don't want to over engineering, but what happens if we need another type of csv for the Adviser model? Since it was added the to_csv to the Adviser model on this extension, the scope of the to_csv, *it's closed for extension*. Can we not use a wrapper on Adviser model instead a module?

File: spec/controllers/admin/reports/registered_adviser_controller_spec.rb (hunk 11 above)
Comment (Tr4pSt3R) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: Thanks for your feedback @tomas-stefano. First I will address the better option:

```
Or even better, you can move the timestamp call to the renderer object and then your acceptance tests cover it.
```

The timestamped filename warranted moving this feature to the controllers where it can be dynamically generated before being passed to the CSV renderer. If the filename was included as part of the acceptance test it will be awkward and unwieldy to test for past, present and future scenarios. 

```
scenario 'permits download as CSV' do
    and_there_exists_registered_advisors
    when_i_visit_the_registered_adviser_page
    and_click_on_download_as_csv_button
    then_i_should_have_a_file_with_all_registered_advisers
  end
```

The scenario above will have to be repeated to cater for different time references as previously mentioned.

File: app/views/admin/reports/registered_adviser/index.html.erb [hunk 13]
```diff
@@ -0,0 +1,41 @@
+<h2>Registered Advisers</h2>
+
+<p>
+  This report shows all advisers whose FCA status is no still 'Active'.
```
Comment (margOnline) [inline app/views/admin/reports/registered_adviser/index.html.erb]: There's a typo here.

File: spec/controllers/admin/reports/registered_adviser_controller_spec.rb (hunk 11 above)
Comment (tomas-stefano) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: 👍

Comment (margOnline) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: I agree with @tomas-stefano that the filename method belongs in the renderer object.

**Comment by tomas-stefano** [comment]
test

File: spec/controllers/admin/reports/registered_adviser_controller_spec.rb (hunk 11 above)
Comment (Tr4pSt3R) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: @margOnline the filename is an option that is merely passed to the render object by the controller though. Since the filename is first mentioned in the controller, I believe this is the best place to test it. wdyt?

File: spec/controllers/admin/reports/registered_adviser_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    before(:each) do
+      Timecop.freeze(date)
+    end
+
+    after(:each) do
+      Timecop.return
+    end
+
+    context 'yesterday' do
```
Comment (norm) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: That date is no longer yesterday, so this description no longer makes sense.

File: lib/ext/adviser.rb (hunk 12 above)
Comment (Tr4pSt3R) [inline lib/ext/adviser.rb]: @tomas-stefano the include is added outside of the `Adviser` class definitions because when it was added to the actual `Adviser` class in the **mas_rad-core** repo you advised that it should be moved out. 

Here is the changes that was proposed to the **mad_rad-core** repo (https://github.com/moneyadviceservice/mas-rad_core/pull/163/files). 

If we use a plain old ruby class to wrap the CSV functionality in what way could it open up the `Adviser` model for further addition of CSV-rendering implementations?

Is this proposal which you suggested the best way to go about this?

```ruby
class AdvisersListCsv
   def initialize(advisers)
      @advisers = advisers
   end

   def self.to_csv(_ = {})
     CSV.generate do |csv|
        csv << ['Ref. Number', 'Name', 'Firm']
 
        @advisers.each do |adviser|
          csv << [adviser.reference_number, adviser.name, adviser.firm.registered_name]
        end
     end
   end
end
```

Comment (willhall88) [inline lib/ext/adviser.rb]: perhaps we could avoid constraining the `.to_csv` method entirely and use a custom method instead (`generate_registered_advisers_csv`)?

Comment (tomas-stefano) [inline lib/ext/adviser.rb]: With the wrapper you manipulate (or you can call extend in a gross way) *without modifying the model*. This opens ways to add more features without the *need to modify the model or touch the to_csv method*.  Imho, this respects the open/closed principle. There is a link on the wiki page on the PR review that @margOnline created that would be interested for you to read: https://maswiki.valiantyscloud.net/display/NewDev/Development+workflow
It is called "Checklist for reviewing code" that there is a small part on open closed principle.

Comment (willhall88) [inline lib/ext/adviser.rb]: or are we able to use the `class_attribute` from **rad**, to extend the Adviser class in **mas_rad-core**? http://api.rubyonrails.org/classes/Class.html#method-i-class_attribute

File: spec/controllers/admin/reports/registered_adviser_controller_spec.rb (hunk 14 above)
Comment (Tr4pSt3R) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: ♻️  @norm && @willhall88 this has been fixed here: https://github.com/moneyadviceservice/rad/pull/406/commits/5e40bb7d84f3a5b24f1f6f41bfd070ff95b970e2

File: lib/ext/adviser.rb (hunk 12 above)
Comment (Tr4pSt3R) [inline lib/ext/adviser.rb]: @tomas-stefano @willhall88 

My question is... what's the best way to go about this? There has been varying expressions of opinions but there's no definite route to moving this implementation forward.

Comment (willhall88) [inline lib/ext/adviser.rb]: @Tr4pSt3R @tomas-stefano I think looking at the current implementation of `.to_csv`, we've actually made it very custom to creating a specific csv file, which then constrains any further extension. IMO that perhaps means that we should not be extending the `.to_csv` method but renaming it as a new custom method.

File: spec/controllers/admin/reports/registered_adviser_controller_spec.rb (hunk 14 above)
Comment (Tr4pSt3R) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: ♻️ @norm && @willhall88 : how about this rather terse context definitions here https://github.com/moneyadviceservice/rad/pull/406/commits/2e87d92add386ddb66f7b06d4df3b1e4a04491da. Do they communicate the context of these tests well enough or is there further room for improvement?

Comment (norm) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: I disagree — the problem is that as a human "yesterday" has a very specific meaning. By setting another forced date in the tests, you have done nothing to remove the confusion that "yesterday" in this context is not "yesterday".

IMO you only need two tests, with very different arbitrary dates (one to prove the filename contains the date, the second to prove that it wasn't hardcoded to that date but is actually dependent on the system date at the time code is executed). Doing yesterday/today/tomorrow is cute but ultimately confusing once enough time has passed.

File: lib/ext/adviser.rb (hunk 12 above)
Comment (tomas-stefano) [inline lib/ext/adviser.rb]: My only opinion was about the use of AdvisersListCsv rather than the module one. The modules are good when it is possible to extend and it is not the case (as I explained above). The problem with just renaming is that than you have a lack of some sort of interface or the duck-typing is lost.

Comment (willhall88) [inline lib/ext/adviser.rb]: resulting in something like this in the controller.. 
```
def index
  @registered_advisers = Adviser.all.page(params[:page])
  respond_to do |format|
    format.html
    format.csv { send_data Adviser.generate_registered_advisers_csv }
  end
end
```

File: spec/controllers/admin/reports/registered_adviser_controller_spec.rb [hunk 15]
```diff
@@ -8,23 +8,23 @@
       Timecop.return
     end
 
-    context 'yesterday' do
+    context 'past date' do
```
Comment (willhall88) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: Looking at this again now. All three tests (past, present, future) are identical and we should strip this down to just one test

Comment (willhall88) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: wdyt @Tr4pSt3R ?

File: app/controllers/admin/reports/registered_adviser_controller.rb [hunk 16]
```diff
@@ -0,0 +1,20 @@
+module Admin
+  module Reports
+    class RegisteredAdviserController < Admin::ApplicationController
+      respond_to :html, :csv
+
+      def index
+        @registered_advisers = Adviser.all.page(params[:page])
+        @registered_advisers_list = AdviserListCsv.new(Adviser.all)
```
Comment (tomas-stefano) [inline app/controllers/admin/reports/registered_adviser_controller.rb]: Can we pass on initialize, the @registered_advisers instead? Will be less hit on the DB.

File: config/application.rb [hunk 17]
```diff
@@ -1,6 +1,7 @@
 require File.expand_path('../boot', __FILE__)
 
 # Pick the frameworks you want:
+require 'csv'
```
Comment (tomas-stefano) [inline config/application.rb]: is this require necessary? Is Rails not loading the csv? And if it is necessary, can we move the renderer initializer?

File: config/application.rb [hunk 18]
```diff
@@ -9,6 +10,8 @@
 require "sprockets/railtie"
 # require "rails/test_unit/railtie"
 
+require_relative '../lib/ext/adviser_list_csv'
```
Comment (tomas-stefano) [inline config/application.rb]: Can we move this require to the class that uses the class?

File: spec/controllers/admin/reports/registered_adviser_controller_spec.rb (hunk 11 above)
Comment (tomas-stefano) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: @Tr4pSt3R  what it is the decision on this?

Comment (Tr4pSt3R) [inline spec/controllers/admin/reports/registered_adviser_controller_spec.rb]: I discussed this with @margOnline and she's happy to keep the tests where they are and as a private method in the controller where it originates from.

File: config/routes.rb [hunk 19]
```diff
@@ -82,6 +82,7 @@
         end
       end
       resource :inactive_adviser, only: [:show]
+      resources :registered_adviser, only: [:index]
```
Comment (willhall88) [inline config/routes.rb]: this can just be `resource` rather than a multiple right?

Comment (Tr4pSt3R) [inline config/routes.rb]: Yeah, it breaks the tests when you change from resources to resource.

**Comment by willhall88** [comment]
👍  I'm happy with this

File: spec/features/admin/export_advisers_list_spec.rb [hunk 20]
```diff
@@ -0,0 +1,96 @@
+# Feature: Advisors list export
+#  As the person maintaining the RAD database.
+#  In order to get rid of unwanted advisors.
+#  I need to be able to cross reference a large list (4k) of unauthorised advisers against records in the RAD database.
+#  The list needs to be cleared up so customers do not get sent to unapproved advisers and RAD maintain it's credibility.
+#
```
Comment (margOnline) [inline spec/features/admin/export_advisers_list_spec.rb]: Does this comment really need to be included?

File: config/routes.rb (hunk 19 above)
Comment (Tr4pSt3R) [inline config/routes.rb]: @willhall88 I'm using resources because the `index` action is a plural resource that will handle generic incoming requests. This is contrary to default actions such as `show`, `new`, `create`, `destroy` and ``update` which works on single resources.

**Comment by margOnline** [comment]
👍

File: spec/features/admin/export_advisers_list_spec.rb (hunk 20 above)
Comment (Tr4pSt3R) [inline spec/features/admin/export_advisers_list_spec.rb]: There's no fast rules about this but I like to include the direct TP story details within the feature itself so that it retains the original business requirement. 

I think they should remain to communicate the rationale behind this feature.

Comment (margOnline) [inline spec/features/admin/export_advisers_list_spec.rb]: I'm not sure the rationale for the feature should be included in the code, but as we have no standards, your call. We can discuss in next week's meeting re Developer Processes/Workflow.

**Comment by tomas-stefano** [comment]
👍 🚢
